vectors/no_data_vec.rs

```diff
@@ -1,359 +1,178 @@
 use std::marker::PhantomData;
 use std::ops::Index;
 use nondet::{
-        Nondet, nondet, nondet_pointer
-    };
-use cvt::{ CVT_assume, CVT_nondet_usize };
-
+    Nondet, nondet, nondet_pointer
+};
+use cvt::CVT_nondet_usize;
 use std::io::Read;
 use anchor_lang::prelude::borsh::maybestd::io::Write;
 use borsh::{BorshDeserialize, BorshSerialize};
 
-////////////////////////////////////////////////////////////////////////
-// Adapted from Kani.
-
-// NoDataVec implements an abstraction of the Vec library which tracks
-// only the length of the vector. It does not contain a backing store
-// which implies that writes only increment the length and all reads
-// return a non-deterministic value.
-////////////////////////////////////////////////////////////////////////
-
-const DEFAULT_CAPACITY: usize = 1073741824;
-const MAX_MALLOC_SIZE: usize = 18014398509481984;
+const MAX_CAPACITY: usize = usize::MAX;
 
 macro_rules! assert {
     ($cond:expr) => {{ /*cvt::CVT_assert($cond)*/}};
 }
 
-// We have observed that resizing might cause collapses in the pointer analysis.
-// So we have this macro to enable/disable resizing.
-macro_rules! allow_resizing {
-    () => { false }
-}
-
-
-// The NoDataVec structure here models the length and the capacity.
 pub struct NoDataVec<T> {
     len: usize,
     capacity: usize,
-    // We use a _marker variable since we want the Vector to be generic over type
-    // T. It is a zero-sized type which is used to mark things such that they act
-    // like they own a T.
-    _marker: PhantomData<T>,
+    _marker: PhantomData<T>
 }
 
 impl<T: Nondet> NoDataVec<T> {
-    // The standard library NoDataVec implementation calls reserve() to reserve
-    // space for an additional element -> self.reserve(1). However, the
-    // semantics of reserve() are ambiguous. reserve(num) allocates space for
-    // "at least num more elements of the containing type". The operation can
-    // be found in function `grow_amortized()` in raw_vec.rs in the standard
-    // library. The logic for choosing a new value is:
-    // self.cap = max(self.cap * 2, self.len + additional)
-    // We try to implement similar semantics here.
-    #[cfg_attr(feature = "certora-debug", inline(never))]
-    fn grow(&mut self, additional: usize) {
-        let new_len = self.len + additional;
-        let grow_cap = self.capacity * 2;
-        let new_capacity = if new_len > grow_cap { new_len } else { grow_cap };
-
-        if new_capacity > MAX_MALLOC_SIZE {
-            panic!("Malloc failed to allocate enough memory");
+    pub fn new() -> Self {
+        Self {
+            len: 0,
+            capacity: MAX_CAPACITY,
```

Comment:
fixing `capacity` to `MAX_CAPACITY` is not an over-approximation. 
Note that some client could have this weird but legitimate code

```
v := NoDataVec<..>::new()
if (v.capacity < 100) {
   ...
 } else {
   ...
}
```
We would claim that the then branch is always false because `capacity = usize::MAX` 
You need to set `capacity` to usize::nondet and add the assumption that `len < capacity`.
Another alternative I think is to remove `capacity` so a vector is only `len` and `_marker` because `capacity` is only used by the method `capacity`.

Reply:
ok fixed